REGRESSION(282648@main): Speedometer 3.0 test TodoMVC-JavaScript-ES6-Webpack-Complex-DOM was regressed

shallawa opened the pull request:
#### 799cdf9de44ee39ead5468f13300884f4bc43f8e
<pre>
REGRESSION(282648@main): Speedometer 3.0 test TodoMVC-JavaScript-ES6-Webpack-Complex-DOM was regressed
<a href="https://bugs.webkit.org/show_bug.cgi?id=279216">https://bugs.webkit.org/show_bug.cgi?id=279216</a>
<a href="https://rdar.apple.com/134688490">rdar://134688490</a>

Reviewed by Ryosuke Niwa.

If an element, without children and without shadowRoot, is insertedIntoAncestor()
don&apos;t call updateEffectiveTextDirection() since it is expensive call. In this
case just, set the bit setEffectiveTextDirection(). To make this change clean,
add the new method updateEffectiveTextDirectionIfNeeded() and call it from
insertedIntoAncestor().

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::updateEffectiveTextDirectionIfNeeded):
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/Element.h:

Canonical link: <a href="https://commits.webkit.org/284731@main">https://commits.webkit.org/284731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e90334fe60f181cba44398c658ca0daca484c35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70320 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23085 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74409 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21498 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57526 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21338 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55728 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14201 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60624 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36190 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41904 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18059 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19859 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63836 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18403 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76128 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14546 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63460 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14588 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60690 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63387 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15584 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11419 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5051 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45528 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/295 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46602 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47879 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46344 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2749,13 +2749,13 @@ bool Element::hasAutoTextDirectionState() const
 {
     auto textDirectionState = elementTextDirectionState(*this);
-    updateEffectiveTextDirectionState(*this, textDirectionState);
+    updateEffectiveTextDirectionState(*this, textDirectionState, false);
```
Can we add an enum class instead of using a boolean?

shallawa replied:
I changed the fix since the old one seems to be incorrect.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2749,13 +2749,13 @@ bool Element::hasAutoTextDirectionState() const
 {
     auto textDirectionState = parseTextDirectionState(newValue);
-    updateEffectiveTextDirectionState(*this, textDirectionState);
+    updateEffectiveTextDirectionState(*this, textDirectionState, true);
```
Ditto.

shallawa replied:
I changed the fix since the old one seems to be incorrect.

shallawa committed:
REGRESSION(282648@main): Speedometer 3.0 test TodoMVC-JavaScript-ES6-Webpack-Complex-DOM was regressed
https://bugs.webkit.org/show_bug.cgi?id=279216
rdar://134688490

Reviewed by Ryosuke Niwa.

If an element, without children and without shadowRoot, is insertedIntoAncestor()
don't call updateEffectiveTextDirection() since it is expensive call. In this
case just, set the bit setEffectiveTextDirection(). To make this change clean,
add the new method updateEffectiveTextDirectionIfNeeded() and call it from
insertedIntoAncestor().

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::updateEffectiveTextDirectionIfNeeded):
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/Element.h:

Canonical link: https://commits.webkit.org/284731@main
